### PR TITLE
Cpt4 2023v1

### PR DESCRIPTION
**Pull request by m-khitrun**

### CODE REVIEW

File: CPT4/readme.md [hunk 1]
```diff
@@ -7,11 +7,17 @@ Prerequisites:
...
+1. Run 
+```sql
+SELECT devv5.FastRecreateSchema(main_schema_name=>'devv5', include_concept_ancestor=>true, include_deprecated_rels=>true, include_synonyms=>true);
+```
 2. Run load_stage.sql (The pVocabularyDate will be automatically retrieved from the UMLS [SOURCES.MRSMAP.vocabulary_date])
 
-3. Run generic_update:
+3. Run check_stage_tables function (should retrieve NULL):
+   ```sql
+    SELECT * FROM qa_tests.check_stage_tables();
```
Comment (hardhouse) [inline CPT4/readme.md:18]: the check_stage_tables function is built into the generic https://github.com/OHDSI/Vocabulary-v5.0/blob/master/working/generic_update.sql#L7

Comment (m-khitrun) [reply CPT4/readme.md]: Fixed. Thank you

File: CPT4/vocab_specific_QA.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+--- Use this counts for analysis and renew respective numbers in https://github.com/OHDSI/Vocabulary-v5.0/wiki/Known-Issues-in-Vocabularies
+
+WITH concepts_in_hierarchy AS (SELECT DISTINCT cr.concept_id_1
+                                FROM concept_relationship cr
+                                        JOIN concept c ON cr.concept_id_1 = c.concept_id
+                                        JOIN concept c1 ON cr.concept_id_2 = c1.concept_id
+                               WHERE c.vocabulary_id = 'CPT4'
+                                 AND c.concept_class_id = 'CPT4'
+                                 AND c1.vocabulary_id IN ('SNOMED', 'OMOP Extension')
+                                 AND cr.relationship_id IN ('Is a', 'Maps to')
```
Comment (TinyRickC137) [inline CPT4/vocab_specific_QA.sql:12]: Do we call it a hierarchy? "Is a" combined with "Maps to" may potentially create ambiguity. I suggest adding 
..retrieves the number of concepts in hierarchy through either Is a or Maps to relationships...

Comment (m-khitrun) [reply CPT4/vocab_specific_QA.sql:12]: Fixed

Comment (TinyRickC137) [reply CPT4/vocab_specific_QA.sql:12]: @Alexdavv disagreed in email

Comment (m-khitrun) [reply CPT4/vocab_specific_QA.sql:16]: Changed the QA logic a bit. Have a look, please.

File: CPT4/vocab_specific_QA.sql [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+SELECT 'concepts_not_in_hierarchy' AS flag,
+       COUNT(concept_id)
+FROM concepts_not_in_hierarchy
+;
+
+-- 2. In CPT4 concepts may migrate between categories and acquire new concept_code. These changes are made in source.
+--- According to the existing logic both concepts remain standard and valid with the only difference of added "(Deprecated)" to the old concept_name
+--- These semantic duplicates should be mapped to concepts with new concept_codes and destandardized.
+SELECT *
```
Comment (TinyRickC137) [inline CPT4/vocab_specific_QA.sql:56]: @m-khitrun could you please specify the fields here? 
At least dates, vocabulary_ids, and standard_concept are not needed here.

File: CPT4/vocab_specific_QA.sql [hunk 4]
```diff
@@ -0,0 +1,65 @@
+-- 1. Our aim is to build a unique hierarchy of Procedures with CPT4 embedded in SNOMED/OMOP Ext hierarchy.
+--- The script below retrieves following data:
+---- 'concepts_in_hierarchy' - the number of concepts in hierarchy through 'Is a' relationships;
+---- 'mapped_concepts' - the number of mapped to SNOMED/OMOP Ext. concepts;
+---- 'concepts_not_in_hierarchy' - the number of concepts that are not yet in hierarchy.
+--- These counts don't include classification concepts and modifiers, and also concepts mapped to other vocabularies.
+--- Use this counts for analysis and renew respective numbers in https://github.com/OHDSI/Vocabulary-v5.0/wiki/Known-Issues-in-Vocabularies
```
Comment (TinyRickC137) [inline CPT4/vocab_specific_QA.sql:7]: I understand, that this is postgeneric check. But can we somehow include previous numbers here to see the dynamics? They are posted on github, should we hardcode them here? Change it to pregeneric checks?
@Alexdavv @m-khitrun thoughts?

File: CPT4/vocab_specific_QA.sql (hunk 3 above)
Comment (m-khitrun) [reply CPT4/vocab_specific_QA.sql]: Done

File: CPT4/vocab_specific_QA.sql (hunk 4 above)
Comment (m-khitrun) [reply CPT4/vocab_specific_QA.sql:7]: We can add the comparison with devv5 as an additional column, like we do in manual checks after generic.

**Review by TinyRickC137 — APPROVED** [review]

**Review by hardhouse — APPROVED** [review]

**Review by TinyRickC137 — APPROVED** [review]

**Pull request merged**